[SYCL][Graph] Update doc for UR PR moving reset commands to a dedicated cmd-list (#357)
reble/llvm: sycl/doc/design/CommandGraph.md

returned is blocking on host.

In order to achieve the expected UR command-buffer enqueue semantics with Level
Zero, the adapter implementation needs extra commands.

> * Prefix - Commands added **before** the relevant workload.
> * Suffix - Commands added **after** the relevant workload.

Reviewer: To try keep the terminology consistent throughout
```suggestion
* Prefix - Commands added **before** the graph workload.
* Suffix - Commands added **after** the graph workload.
```

These extra commands operate on L0 event synchronisation primitives, 
used by the command-list to interact with the external UR wait-list 
and UR return event required for the enqueue interface.
Unlike the graph workload (i.e. commands needed to perform the graph workload) 
the external UR wait-list and UR return event are submission dependent, 
which mean they can change from one submission to the next.

For performance concerns, the command-list that will execute the graph
workload is made only once (during the command-buffer finalization stage).
This allows the adapter to save time when submitting the command-buffer, 
by executing only this command-list (i.e. without enqueuing any commands 
of the graph workload).

#### Prefix

The prefix's commands aim to: 

> 1. Handle the the list on events to wait on, which is passed by the Run-Time

Reviewer: ```suggestion
1. Handle the the list on events to wait on, which is passed by the runtime
```

> when the command-buffer enqueue function is called.

Reviewer: ```suggestion
when the UR command-buffer enqueue function is called.
```

> As mentionned above, this list of events changes from one submission 

Reviewer: ```suggestion
As mentioned above, this list of events changes from one submission 
```

to the next.

> Consequently, managing this mutable command in the graph-workload 

Reviewer: ```suggestion
Consequently, managing this mutable dependency in the graph-workload 
```

command-list implies rebuilding the command-list for each submission 
(note that this can change with mutable command-list).
To avoid the signifiant time penalty of rebuilding this potentially large 

> command-list each time, we prefer to add the extra command handling the

Reviewer: ```suggestion
command-list each time, we prefer to add an extra command handling the
```

wait list into another command-list (*wait command-list*). 
This command-list consists of a single L0 command: a barrier that waits for 

>  dependencies passed by the wait-list and signals a signal 

Reviewer: ```suggestion
dependencies passed by the wait-list and signals a signal 
```

called *WaitEvent* when the barrier is complete.
This *WaitEvent* is defined in the `ur_exp_command_buffer_handle_t` class.
In the front of the graph workload command list, an extra barrier command 
waiting for this event is added (when the command-buffer is created).
This ensures that the graph workload does not start running before 
the dependencies to be completed.
The *WaitEvent* event is reset in the suffix.

2. Reset events associated with the command-buffer except the 
*WaitEvent* event.
Indeed, L0 events needs to be explicitly reset by an API call 
(L0 command in our case).

> Since a command-buffer is expected to be submitted multiple times. 
> We need to ensure that L0 events associated with graph commands have not

Reviewer: ```suggestion
Since a command-buffer is expected to be submitted multiple times,
we need to ensure that L0 events associated with graph commands have not
```

been signaled by a previous execution. These events are therefore reset to the
non-signaled state before running the graph-workload command-list. Note
that this reset is performed in the prefix and not in the suffix to avoid
additional synchronization w.r.t profiling data extraction.
We use a new command list (*reset command-list*) for performance concerns. 
Indeed:

>    * This allows the `waitEvent` to be signaled directly on the host if 

Reviewer: This is how it's referred to previously
```suggestion
   * This allows the *WaitEvent* to be signaled directly on the host if 
```

   the waiting list is empty, thus avoiding the need to submit a command list.

>    * Enqueuing a reset ZE command for all events in the command-buffer is time

Reviewer: Haven't really used the "ZE" terminology elsewhere
```suggestion
   * Enqueuing a reset L0 command for all events in the command-buffer is time
```

   consumming, especially for large graphs.
   However, this task is not needed for every submission, but only once, when the
   command-buffer is fixed, i.e. when the command-buffer is finalized. The
   decorelation between the reset command-list and the wait command-list allow us to
   create and enqueue the reset commands when finalizing the command-buffer, 
   and only create the wait command-list at submission.

This command list is consist of a reset command for each of the graph commands 
and another reset command for resetting the signal we use to signal the completion 
of the graph workload. This signal is called *SignalEvent* and is defined in 
in the `ur_exp_command_buffer_handle_t` class.

#### Suffix

The suffix's commands aim to:
1) Handle the completion of the graph workload and signal 
an UR return event.
Thus, at the end of the graph workload command-list a command, which 
signals the *SignalEvent*, is added (when the command-buffer is finalized).
In an additional command-list (*signal command-list*), a barrier waiting for 
this event is also added.
This barrier signals, in turn, the UR return event that has be defined by 

> the Run-Time layer when calling the `urCommandBufferEnqueueExp` function.

Reviewer: ```suggestion
the runtime layer when calling the `urCommandBufferEnqueueExp` function.
```

non-signaled state before running the actual graph associated commands. Note
that this reset is performed in the prefix and not in the suffix to avoid
additional synchronization w.r.t profiling data extraction.
Note also that the prefix has been split into two command-lists for
performance concerns.
1) This allows the `waitEvent` to be signaled directly on the host if 
the waiting list is empty, thus avoiding the need to submit a cmd list.
2) Enqueuing a reset ZE command for all events in the command-buffer is time
consumming, especially for large graphs.
However, this task is not needed for every submission, but only once, when the
command-buffer is fixed, i.e. when the command-buffer is finalized. The

> decorelation between the reset cmd-list and the wait command-list allow us to

Reviewer: ```suggestion
decorellation between the reset cmd-list and the wait command-list allow us to
```
